File: cpp/example_1_spin_rwa.cpp
/***************************************************************
* Author: J.P.G. van Dijk
***************************************************************/

// Performance/accuracy settings
//#define MKL					(define at project level)
//#define DOUBLE_PRECISION		(define at project level)

// Includes
#define _USE_MATH_DEFINES
#include <math.h>
#include <iostream>
#include "spine/simulate.h"
#include "spine/fidelity.h"
#include "spine/math.h"
#include "spine/solvers/solver_taylor.h"
#include "spine/systems/system_1_spin_rwa.h"

// Namespaces
using namespace std;
using namespace spine::math;

// Function prototypes
bool inHamiltonian(complex * H, realnum * dt);
void outOperation(complex * U);

// Private global variables/constants
static spine::systems::system_1_spin_rwa spin_system = spine::systems::system_1_spin_rwa();
static const realnum dt = (realnum) 50e-12;
static int Nsim, nsim;

int main(void)
{
	// Create the system
	spin_system.setMicrowaveFrequency((realnum)(2 * M_PI * 1e9 * 0));	// Set to zero for nicer plot
	spin_system.setLarmorFrequency((realnum)(2 * M_PI * 1e9 * 0));		// Set to zero for nicer plot
	spin_system.setRabiFrequency((realnum)(2 * M_PI * 1e6));

	// Generate the driving signal
	realnum tpi = (realnum)M_PI / spin_system.getRabiFrequency();
	int Npi = (int)(tpi / dt);

	// Simulate the system
	nsim = 0;
	Nsim = Npi;
#ifdef PLOT
	spin_system.plotSetup(Nsim, 0, tpi);
#endif
	cout << "Simulating..." << endl;
	spine::simulate(spin_system.getDimension(), inHamiltonian, outOperation, spine::solvers::solver_taylor);
	cout << "finished!" << endl;
#ifdef PLOT
	Plot::run();
#endif
}

bool inHamiltonian(complex * H, realnum * dt)
{
	if (nsim < Nsim)
	{
		// Set the signal at this time instance
		spin_system.setMicrowaveAmplitude(1);
		if (nsim < (Nsim / 2))
			spin_system.setMicrowavePhase(0);
		else
			spin_system.setMicrowavePhase((realnum)(M_PI / 2.0));

		// Update the Hamiltonian accordingly
		spin_system.updateHamiltonian(H);

		// Provide the current timestep, and continue the simulation
		*dt = ::dt;
		return true;
	}
	return false;
}

void outOperation(complex * U)
{
	// Plot every Nplot-th point
	const unsigned int Nplot = 5;

	// Print the measurement probability
#ifdef PLOT
	spin_system.plotAddU(U, nsim * dt);
#endif
	if ((nsim % Nplot) == 0)
	{
		realnum P = abs2(U[0]);
		cout << P << endl;
#ifdef PLOT
		spin_system.plot();
#endif
	}

	// Continue the simulation
	nsim++;
	if (nsim == Nsim)
	{
		// The ideal X, pi / 2 rotation
		complex Ux[4] = { {cos(M_PI / 4), 0}, {sin(0) * sin(M_PI / 4), -cos(0) * sin(M_PI / 4)}, {-sin(0) * sin(M_PI / 4), -cos(0) * sin(M_PI / 4)}, {cos(M_PI / 4), 0} };
		complex Uy[4] = { {cos(M_PI / 4), 0}, {sin(M_PI / 2) * sin(M_PI / 4), -cos(M_PI / 2) * sin(M_PI / 4)}, {-sin(M_PI / 2) * sin(M_PI / 4), -cos(M_PI / 2) * sin(M_PI / 4)}, {cos(M_PI / 4), 0} };
		complex Uideal[4];
		multiply(2, Uy, Ux, Uideal);

		// Last point, print the fidelity
		realnum F = spine::fidelity(spin_system.getDimension(), U, Uideal);
		cout << "Fidelity: " << F << endl;
	}
}

File: matlab/example_1_spin_rwa.m
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% Author: J.P.G. van Dijk                                                 %
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

% Cleanup
clear all;
close all;
clc;

% Global variables
global nsim;
global Nsim;
global system;
global microwave_amplitude;
global microwave_phase;
global dt;
global probability;

% Create the system
system = spine.systems.system_1_spin_rwa();
system.setMicrowaveFrequency(2 * pi * 1e9);
system.setLarmorFrequency(2 * pi * 1e9);
system.setRabiFrequency(2 * pi * 1e6);

% Generate the driving signal
dt = 50e-12;
tpi = pi / system.getRabiFrequency();
Npi = tpi / dt;
t = (1:Npi) * dt;
microwave_amplitude = ones(1, Npi);
microwave_phase = [zeros(1, Npi/2), pi/2*ones(1, Npi/2)];

% Simulate the system
nsim = 0;
Nsim = Npi;
spine.simulate(system.getDimension(), @inHamiltonian, @outOperation, @spine.solvers.solver_expm);

% Plot the result
figure();
plot(t, probability);

function [run, H, timestep] = inHamiltonian()
    global nsim;
    global Nsim;
    global system;
    global microwave_amplitude;
    global microwave_phase;
    global dt;

    if (nsim < Nsim)
        
        % Set the signal at this time instance
        system.setMicrowaveAmplitude(microwave_amplitude(nsim + 1));
        system.setMicrowavePhase(microwave_phase(nsim + 1));
        
        % Update the Hamiltonian accordingly
        H = system.updateHamiltonian();
        
        % Provide the current timestep, and continue the simulation
        timestep = dt;
        run = true;
        
    else
        run = false;
        H = [];
        timestep = [];
    end
    
end

function outOperation(U)
    global nsim;
    global Nsim;
    global system;
    global probability;
    
    % Plot every Nplot-th point
    Nplot = 5;
    
    % Store the measurement probability
    probability(nsim + 1) = abs(U(1,1))^2;
    
    % Show the rotation in the Bloch sphere
    if (mod(nsim, Nplot) == 0)
        system.plot(U, 0, 1);  % t = 0, as we are already in the rotating frame!
        pause(0.001);
    end
    
    % Continue the simulation
    nsim = nsim + 1;
    if (nsim == Nsim)
        
        % The ideal X, pi/2 rotation
        Ux = [cos(pi/4), -1i * (cos(0) + 1i * sin(0)) * sin(pi/4);
              -1i * (cos(0) - 1i * sin(0)) * sin(pi/4), cos(pi/4)];
        Uy = [cos(pi/4), -1i * (cos(pi/2) + 1i * sin(pi/2)) * sin(pi/4);
              -1i * (cos(pi/2) - 1i * sin(pi/2)) * sin(pi/4), cos(pi/4)];
        Uideal = Uy * Ux;
        
        % Last point, print the fidelity and unitary matrix
        F = spine.fidelity(system.getDimension(), U, Uideal);
        disp(F);
        disp(U);
        
    end
    
end
